contrib/babelfishpg_tsql/src/cursor.c

```diff
@@ -1468,6 +1471,42 @@ execute_sp_cursoropen_common(int *stmt_handle, int *cursor_handle, const char *s
 
 	if (prepare)
 	{
+		/*
+		 * This entire block is to parse the statement by antlr and use the resultant
+		 * statement to sent to Postgres cursor execution. This is necessary in some use case,
+		 * for example when we have PostgreSQL reversed keywords in query which is valid in TSQL.
+		 * Antlr parser will add the quotes at necessary places in query so that Postgres engine
+		 * can resolve this query correctly.
+		 */
+		if (stmt)
+		{
+			/* Copy the original statement, because we don't want to use the resultant query in every case. */
+			stmt_copy = pstrdup(stmt);
+			/* Send to antlr parser */
+			func = pltsql_compile_inline(stmt_copy, NULL);
+
+			/*
+			 * Check the node list of type PLtsql_stmt_type. Cursor only support single statement.
+			 * If there are more than 1 statement we will through the error. func->action-body
+			 * returned by pltsql_compile_inline contains two default nodes, PLTSQL_STMT_INIT being first
+			 * and PLTSQL_STMT_RETURN being last. So total number of nodes should be 3 for cursor. Actual
+			 * query statement will be at second position of type PLTSQL_STMT_EXECSQL.
+
+			 * This is defensive code, where we only reassign the stmt variable to parsed query,
+			 * if the cmd_type is PLTSQL_STMT_EXECSQL. There might be other types of cmd_type like
+			 * PLTSQL_STMT_EXECSQL (for procedures), for them we will keep the old behavior.
+			 */
+			if((( (PLtsql_stmt *) lsecond(func->action->body))->cmd_type == PLTSQL_STMT_EXECSQL)
+				&& list_length(func->action->body) == 3)
+			{
+				parse_result = (PLtsql_stmt_execsql *) lsecond(func->action->body);
+				stmt = pstrdup(parse_result->sqlstmt->query);
```

Comment:
1. stmt is declared as a const char *. It's best practice not to reassign a const pointer. We can either think of changing the const declaration or by having another local tsql_stmt for this case.

2. When overriding a pointer that refers to palloc'ed memory, we should ensure the original memory is freed before assignment to prevent memory leaks. Here since it is const, I assume we might get error while freeing it. If we introduce a new variable for the rewritten statement, this problem is avoided.